app/public/cantusdata/helpers/mei_processing/mei_tokenizer.py

```diff
@@ -85,67 +64,176 @@ def __init__(self, mei_file: str, min_ngram: int, max_ngram: int) -> None:
         self.min_ngram = min_ngram
         self.max_ngram = max_ngram
 
-    def get_neume_ngram_docs(self) -> List[NeumeNgramDocument]:
+    @property
+    def flattened_neumes(self) -> List[Neume]:
         """
-        Generate neume-level documents for search, containing
-        n-grams of neume names.
+        Flatten the neumes contained in the syllables of the MEI file.
 
-        :return: A list of dictionaries containing the n-grams
-            of neume names.
+        :return: A list of neumes.
         """
-        neumes_sequence: List[Neume] = []
+        neumes: List[Neume] = []
         for syllable in self.syllables:
```

Comment:
I understand that you are removing empty syllables and neumes from MEI files before parsing, but what if the MEI file itslef is empty or has no neumes? Should this function be Optional[List[Neume]]?

Reply:
Ah, ok .. yeah, let me check that

Reply:
Ok, I think this actually is the correct type, but it has taken a bit of research to figure out why.

Basically, `Optional[List[Neume]]` is the same as saying `neumes` could be either `List[Neume]` or `None`. But that raises a type error because `neumes` won't ever actually be `None`, because at the very least (in the case with no syllables or no neumes in the MEI file) it will be `[]`.

For similar reasons, `List[Optional[Neume]]` doesn't work, because we'll never have `None` in the list, we just may have a list of cardinality 0. 

[This post](https://stackoverflow.com/questions/76881483/whats-the-correct-way-to-type-hint-an-empty-list-as-a-literal-in-python) led me to looking into the type annotation `Never`, so I tried something here like `Union[List[Neume], List[Never]]`, but that doesn't work because `Never`, as the name suggest, means there won't ever be something added but some times/most of the time, there will be something added. Tbh, I'm kinda confused about `Never` so this paragraph might not be exactly right.

So, yeah, I wonder if `List[Neume]` is correct because even `[]` is a list of neumes (it's just a list of 0 neumes). But, there may be another, better annotation, I just can't figure out what it is.

Reply:
If you're assigning it to `[]` then it will always be a `List`. 

> Tbh, I'm kinda confused about Never so this paragraph might not be exactly right.

To understand `Never` I would recommend learning about Functional and Statically Typed languages like Haskell or Elm. `Never` means you can use the type system to guarantee that something will never exist. If you try to do something with a value that does exist, it will be an error that you can catch at the type / compiler level, rather than at runtime. A type of `list[Never]` means that it will be a list that can't hold any values. 

If you think of an `Int` as being something like `Int 1`, `Int 2`, etc. (e.g., a type with a value of 1 or 2), and a `String` as something like `String "abc"`, `String "hello"`, then these are both types that have values. Likewise, `List[Int]` or `List[String]` are lists that accept ints or strings, i.e., `[Int 1, Int 2, Int 3]`. Likewise, even `None` is a value, so `List[Optional[Int]]` can mean `[None, Int 1, None]`. (You know it's a value because it has an index in the list.)

A `List[Never]` can never hold anything, not even `None`. So how do you add a value to a list if that value can't be expressed? You can't. Thus `x: List[Never] = []`, and if you try to `x.append(1)` the type system should flag it and say that `1` is not of type `Never`. Since you can never instantiate a value to satisfy the type system, `x` will always be an empty list. 

It's probably not what you want in this case, though. I thought I would just chime in with that. Learning a Functional Programming language is perhaps the most useful thing I did to help me write better code.